lib/improver/psychrometric_calculations/psychrometric_calculations.py

```diff
@@ -909,28 +919,114 @@ def fill_in_sea_points(
+           points where there is a valid snow falling level which is below
+           the maximum orography in the region arround that point. This
+           helps us avoid spreading very high snow falling levels across
+           areas where we had missing data.
+        2. Fill in in any points where we have not been able to interpolate
+           as there is not enough data (e.g at the corners of the domain),
+           using a nearest neighbour algorithm.
+        3. Check whether dispite our efforts we have still filled some
+           points with snow falling levels above the orography. In these
+           cases set them to the height of orography.
+
+        We then return the filled in array, which hopefully has no more
+        missing data.
 
         Args:
             snow_level_data(numpy.array):
                 The snow falling level array, filled with values for points
                 whose wet bulb temperature integral crossed the theshold.
+            max_in_nbhood_orog(numpy.array):
+                The array containing maximum of the orography field in
+                a given radius.
+            orog_data(numpy.data):
+                The array containing the orography data.
         Returns:
-            snow_level_data_updated(numpy.array):
+            snow_filled(numpy.array):
                 The snow falling level array with missing data filled by
                 horizontal interpolation.
         """
         # Interpolate linearly across the remaining points
-        points = np.where(np.isfinite(snow_level_data))
-        values = snow_level_data[points]
-        ynum, xnum = snow_level_data.shape
-        (y_points, x_points) = np.mgrid[0:ynum, 0:xnum]
-        snow_level_data_updated = griddata(
-            points, values, (y_points, x_points), method='linear')
-        return snow_level_data_updated
+        index = (snow_level_data <= max_in_nbhood_orog)
+        snow_filled = snow_level_data
+        if np.any(index):
+            ynum, xnum = snow_level_data.shape
+            (y_points, x_points) = np.mgrid[0:ynum, 0:xnum]
+            points = np.where(index)
+            values = snow_level_data[points]
+            snow_level_data_updated = griddata(
+                points, values, (y_points, x_points), method='linear')
+            snow_filled = snow_level_data_updated
+            # Fill in any remaining missing points using nearest neighbour.
+            # This normallly only impact points at the corners of the domain,
+            # where the linear fit doesn't reach.
+            index = (snow_level_data_updated <= max_in_nbhood_orog)
+            if np.any(index):
+                points = np.where(index)
+                values = snow_level_data_updated[points]
+                snow_level_data_updated_2 = griddata(
+                    points, values, (y_points, x_points), method='nearest')
+                snow_filled = snow_level_data_updated_2
```

Comment:
I need to sort out the indexing in this section anyway to get rid of the warnings. I'll neaten this up at the same time now I've got unit tests that I know will capture mistakes if I mess it up.

Reply:
The reason snow_filled is set in so many places is so that everything is so that we return something even if it doesn't get through the if statements. I've changed this around a bit now and got rid of the warnings, see what you think. I may have made it harder to follow and I expect I need better variable names rather than keep reusing `index`

Reply:
Looks fine, but I agree the repeated use of index could possibly be improved.